design/ip-address-management-for-networkdata.md

```diff
@@ -0,0 +1,328 @@
+or kubeadm control plane referencing it use it, effectively sharing it among
+multiple machine deployments.
+
+#### Story 3
+
+As a user deploying a target Kubernetes cluster, I want to keep the allocation
+predictability offered by the current feature, i.e. the ip address is the
+start address + the index of the Metal3Data, when I use a pool for a single
+machine deployment. This is to ensure that the current proposal adds a feature
+without removing any existing feature and is fully backward compatible.
+
+### Implementation Details/Notes/Constraints
+
+- This proposal should be fully backwards compatible and not modify any existing
+  behaviour.
+- When shared among multiple machine deployment, the allocation would be
+  random as much as possible to avoid conflicts and re-use as much as possible.
+- If not needed, this feature should not require the user to modify anything
+  existing in their CRs
+
+### Risks and Mitigations
+
+This adds some complexity to the Metal3 system. Proper documentation could solve
+the issue, and for users not needing this feature, no change would be required.
+
+## Design Details
+
+
+A Metal3IPPool object would be created :
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3IPPool
+metadata:
+  name: pool-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Cluster
+    name: cluster-1
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Data
+    name: metal3data-1
+spec:
+  pools:
+    - start: 192.168.0.10
+      end: 192.168.0.15
+      subnet: 192.168.0.0/24
+      step: 1
+    - start: 192.168.0.20
+      end: 192.168.0.25
+      subnet: 192.168.0.0/24
+status:
+  lastUpdated: "2020-04-02T06:36:09Z"
+  allocations:
+    "192.168.0.11": "metal3data-1"
+  addresses:
+    "metal3data-1": "pool-1-192-168-0-11"
+  error: false
+  errorMessage: ""
+```
+
+In the *spec* of *Metal3IPPool*, there would be a *pools* list, that would
+contain a list of IP address pools, with the *start* and *end* attributes
+giving the start and end ip addresses of the pool, the *subnet* allowing to
+verify that the allocated IP is in the pool and from which the start and end ip
+addresses can be inferred. The *step* item gives the interval between allocated
+IP addresses. Specifying single ip addresses can be achieved by setting the
+start and end ip address to that single ip address.
+
+The *status* would contain a *lastUpdated* field with the timestamp of the last
+update and and *error* boolean that is set to true if an error during allocation
+happens. In that case, the *errorMessage* will contain the text of the error.
+The *allocations* map will map the IP address to the *Metal3Data* object it was
+allocated for and the *addresses* will map the *metal3Data* objects with the
+*Metal3IPAddress* objects.
+
+All IP addresses specified in a Metal3IPPool must be from the same subnet.
```

Comment:
good point, that was a mix from the initial proposal with multiple subnets in the pool. However, it makes it much more complicated to use since in the network yaml description, you would also then need to fetch the mask and the gateway if there are different subnets. Hence restricting it to a single subnet. I'll fix this.

Reply:
I could see it being useful to allow multiple subnets to support different failure domains in the switches. Do you think we absolutely need all of the IPs to be on the same subnet?

Reply:
Would you mind expanding your point about failure domains ? If we want IP addresses from multiple subnets, we could have multiple pools and refer to the as many as needed in the networkdata interface configuration. I'm wondering whether multiple subnets in a pool would introduce complexity with regards to the mask and gateway configuration. To be honest I am undecided here. Implementing support for multiple subnets would also cover using a single subnet, so maybe that's the way we should go ?